### PR TITLE
Fixes for <exclude/> support, with tests and docs.

### PR DESCRIPTION
**Pull request by davedawkins**
Documentation on rebasing links
Issues #786 , #785

### CODE REVIEW

File: FSharp.Formatting.sln [hunk 1]
```diff
@@ -83,6 +83,8 @@ Project("{6EC3EE1D-3C4E-46DD-8F32-0CC8E7565705}") = "TestLib1", "tests\FSharp.Ap
 EndProject
 Project("{6EC3EE1D-3C4E-46DD-8F32-0CC8E7565705}") = "TestLib2", "tests\FSharp.ApiDocs.Tests\files\TestLib2\TestLib2.fsproj", "{48EFFECF-ECB0-4DF3-A704-B56AB07557BF}"
 EndProject
+Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "TestLib3", "tests\FSharp.ApiDocs.Tests\files\TestLib3\TestLib3.fsproj", "{52B949AA-A3F7-4894-B713-804BAEB71118}"
```
Comment (nojaf) [inline FSharp.Formatting.sln:86]: `TestLib3` isn't showing up under the right folder for me:
image

File: docs/apidocs.fsx [hunk 2]
```diff
@@ -169,6 +169,63 @@ type GenericClass2<'T>() =
...
+
+If you want to exclude modules or functions from the API docs you can use the `<exclude/>` tag.
+It needs to be set on a separate triple-slashed line, and can either appear on its own or as part
+of an existing `<summary>` (for example, you may wish to hide existing documentation while it's in progress).
+The `<exclude/>` tag can be the first or last line in these cases.
+
+Some examples:
+*)
+
+/// </exclude>
```
Comment (nojaf) [inline docs/apidocs.fsx:183]: Should this not be `<exclude />`? `</exclude>` without opening tag doesn't strike me as valid XML.

File: src/FSharp.Formatting.ApiDocs/GenerateHtml.fs [hunk 3]
```diff
@@ -707,7 +707,7 @@ type HtmlRender(model: ApiDocModel, ?menuTemplateFolder: string) =
             printfn "  Generating %s" outFile
             SimpleTemplating.UseFileAsSimpleTemplate(substitutions, templateOpt, outFile)
 
-        for info in model.EntityInfos do
+        for info in model.EntityInfos |> List.filter (fun info -> not info.Entity.Exclude) do
```
Comment (nojaf) [inline src/FSharp.Formatting.ApiDocs/GenerateHtml.fs:710]: Should we consider having the filtering inside `.EntityInfos` instead?
Based on the usage of this member I think this makes sense.

File: src/FSharp.Formatting.ApiDocs/GenerateModel.fs [hunk 4]
```diff
@@ -2261,12 +2261,30 @@ module internal SymbolReader =
...
+            // 1.
+            //  // Not triple-slash ccomment
+            //  /// <exclude/>
+            //
+            // 2.
+            //  /// <exclude/>
+            //
+            // So, we need to let the 'null' case handle this to extract the <exclude> if it's there
+            //
+            // | null when String.IsNullOrEmpty el.Value ->
```
Comment (nojaf) [inline src/FSharp.Formatting.ApiDocs/GenerateModel.fs:2277]: These last two lines of the comment can be omitted?

File: src/FSharp.Formatting.ApiDocs/GenerateModel.fs [hunk 5]
```diff
@@ -2261,12 +2261,30 @@ module internal SymbolReader =
...
+            //     cmds, ApiDocComment.Empty, None
+
             | null ->
                 // We let through XML comments without a summary tag. It's not clear
                 // why as all XML coming through here should be from F# .XML files
                 // and should have the tag.  It may be legacy of previously processing un-processed
                 // XML in raw F# source.
+                //
+                // 9-Jan-23: See comment above for at least one reason why we pass through here now
```
Comment (nojaf) [inline src/FSharp.Formatting.ApiDocs/GenerateModel.fs:2286]: This journal entry seems a tad excessive. You explained the reasoning not ten lines ago.

File: src/FSharp.Formatting.ApiDocs/GenerateModel.fs-debug [hunk 6]
```diff
@@ -0,0 +1,3157 @@
+namespace rec FSharp.Formatting.ApiDocs
```
Comment (nojaf) [inline src/FSharp.Formatting.ApiDocs/GenerateModel.fs-debug:1]: Should this file exist?

**Review by nojaf — CHANGES_REQUESTED** [review]
This looks like a solid PR. I have some minor nitpicks along the way but I think we can take this in.
At the same time, I'm no expert so @nhirschey do you want to glance over this one as well?

@davedawkins if you update the `RELEASE_NOTES.md` file with your fix and a new version (say `17.2.1`) a new version would be released once we merged this to main.
If you don't need a new version that's fine as well.

File: src/FSharp.Formatting.ApiDocs/GenerateModel.fs-debug (hunk 6 above)
Comment (davedawkins) [reply src/FSharp.Formatting.ApiDocs/GenerateModel.fs-debug:1]: No - will fix

File: src/FSharp.Formatting.ApiDocs/GenerateModel.fs (hunk 5 above)
Comment (davedawkins) [reply src/FSharp.Formatting.ApiDocs/GenerateModel.fs:2286]: Well, it's a single line pointing at the explanation, in response to the question raised in the previous developer's comment "why are we here". How do you want to improve this?

File: src/FSharp.Formatting.ApiDocs/GenerateHtml.fs (hunk 3 above)
Comment (davedawkins) [reply src/FSharp.Formatting.ApiDocs/GenerateHtml.fs:710]: I will take a look at EntifyInfos. My preference was to do it in the logic that calls `checkAccess`, but I believe that's before we've processed the comments.

File: docs/apidocs.fsx (hunk 2 above)
Comment (davedawkins) [reply docs/apidocs.fsx:183]: Yes, a typo. Sorry :-/

File: FSharp.Formatting.sln (hunk 1 above)
Comment (davedawkins) [reply FSharp.Formatting.sln:86]: I will fix this.

File: src/FSharp.Formatting.ApiDocs/GenerateModel.fs (hunk 5 above)
Comment (davedawkins) [reply src/FSharp.Formatting.ApiDocs/GenerateModel.fs:2286]: I removed all the comments. I see your point (to an extent), however: the original comment either needs a response (in context), or just removing. My approach is not to remove comments, but respond to them in a timeline to help out developers down the line.  I think however nothing is lost - if you think it's excessive, others may do too.

**Comment by davedawkins** [comment]
All updates made as requested.

File: src/FSharp.Formatting.ApiDocs/GenerateModel.fs [hunk 7]
```diff
@@ -2261,12 +2261,21 @@ module internal SymbolReader =
             let sum = el.Element(XName.Get "summary")
 
             match sum with
```
Comment (nojaf) [inline src/FSharp.Formatting.ApiDocs/GenerateModel.fs:2263]: The match is looking a bit funny now.
Perhaps go with:
```fsharp
            // sum can be null with null/empty el.Value when an non-"<summary>" XML element appears
            // as the only '///' documentation command:
            //
            // 1.
            //  // Not triple-slash ccomment
            //  /// <exclude/>
            //
            // 2.
            //  /// <exclude/>
            if isNull sum then
                let doc, nsels = readXmlCommentAsHtmlAux false ctx.UrlMap el cmds

                let nsdocs = readNamespaceDocs ctx.UrlMap nsels
                cmds, doc, nsdocs
            elif ctx.MarkdownComments then
                readMarkdownCommentAndCommands ctx sum.Value el cmds
            else
                if sum.Value.Contains("<exclude") then
                    cmds.["exclude"] <- ""

                    printfn
                        "Warning: detected \"<exclude/>\" in text of \"<summary>\" for \"%s\". Please see https://fsprojects.github.io/FSharp.Formatting/apidocs.html#Classic-XML-Doc-Comments"
                        xmlSig

                readXmlCommentAndCommands ctx sum.Value el cmds
```
instead.

**Review by nojaf — APPROVED** [review]
Looks good to me.

File: src/FSharp.Formatting.ApiDocs/GenerateModel.fs (hunk 7 above)
Comment (davedawkins) [reply src/FSharp.Formatting.ApiDocs/GenerateModel.fs]: Agreed, your fix applied

**Comment by nhirschey** [comment]
Thanks so much @davedawkins. @nojaf sure, I should be able take a look in a few hours.

**Comment by davedawkins** [comment]
Please hold off from applying this PR. I think I've found an issue with excluding `type`.

**Comment by davedawkins** [comment]
There's no issue with excluding type as such, but the module will still list an excluded type. Additionally, the search index is listing excluded items, so I'm going to fix both those issues today. Unless requested otherwise, I'm going to check those changes into this branch (and PR).

**Comment by nojaf** [comment]
Hello @davedawkins, thank you so much for your persistence here!

I think having the index changes in this PR is ok. I took a peek at the latest changes and everything is still ok for me.

**Comment by davedawkins** [comment]
@nojaf, no problem. I'm determined to get Sutil's API documented! Thank you for help too.

**Review by nhirschey — APPROVED** [review]

**Comment by nhirschey** [comment]
Wonderful set of fixes @davedawkins. Thank you! It looks good to me as well. 

I went ahead and merged the PR to trigger the release. It should hit the feed soon.

**Pull request merged**